<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>index (bio_io.index)</title><link rel="stylesheet" href="../odoc.css"/><meta charset="utf-8"/><meta name="generator" content="odoc 2.1.1"/><meta name="viewport" content="width=device-width,initial-scale=1.0"/><script src="../highlight.pack.js"></script><script>hljs.initHighlightingOnLoad();</script></head><body class="odoc"><nav class="odoc-nav"><a href="../index.html">Up</a> â€“ bio_io</nav><header class="odoc-preamble"><h1 id="bio_io"><a href="#bio_io" class="anchor"></a>Bio_io</h1><p><a href="https://github.com/mooreryan/bio_io/actions/workflows/build_and_test.yml">
<img
src="https://github.com/mooreryan/bio_io/actions/workflows/build_and_test.yml/badge.svg?branch=main"
alt="Build and test"/> </a> <a href="https://coveralls.io/github/mooreryan/bio_io?branch=main">
<img
src="https://coveralls.io/repos/github/mooreryan/bio_io/badge.svg?branch=main"
alt="Coverage Status"/> </a></p><p><b><a href="Bio_io/index.html">The full API is browsable here</a></b>.</p><p><code>Bio_io</code> is an OCaml library that provides programmatic access to common file formats used in bioinformatics like FASTA files.</p><p>If you have any problems or find any bugs, open an issue on the <a href="https://github.com/mooreryan/bio_io/issues">GitHub page</a>.</p></header><nav class="odoc-toc"><ul><li><a href="#license">License</a></li><li><a href="#quick-start">Quick Start</a><ul><li><a href="#install">Install</a></li><li><a href="#example">Example</a></li></ul></li><li><a href="#overview">Overview</a></li><li><a href="#extending-bio_io">Extending Bio_io</a></li></ul></nav><div class="odoc-content"><h2 id="license"><a href="#license" class="anchor"></a>License</h2><p><a href="https://github.com/mooreryan/bio_io"> <img
src="https://img.shields.io/badge/license-MIT%20or%20Apache%202.0-blue"
alt="license MIT or Apache 2.0"/> </a></p><p>Licensed under the Apache License, Version 2.0 or the MIT license, at your option. This program may not be copied, modified, or distributed except according to those terms.</p><h2 id="quick-start"><a href="#quick-start" class="anchor"></a>Quick Start</h2><h3 id="install"><a href="#install" class="anchor"></a>Install</h3><pre><code>opam install bio_io</code></pre><h3 id="example"><a href="#example" class="anchor"></a>Example</h3><p>Read a FASTA file and print the ID and sequence length for each record.</p><pre><code>open! Base

let fasta_file = &quot;sequences.fasta&quot;

let () =
  (* This open gives you [In_channel] and [Record]. *)
  let open Bio_io.Fasta in
  In_channel.with_file_iter_records fasta_file ~f:(fun record -&gt;
      (* Print the ID and the length of the sequence. *)
      Stdio.printf &quot;%s =&gt; %d\n&quot; (Record.id record) (Record.seq_length record))</code></pre><h2 id="overview"><a href="#overview" class="anchor"></a>Overview</h2><p>The <code>Bio_io</code> library provides input channels that return records.</p><p>For an overview see the <a href="Bio_io/Record_in_channel/module-type-S/index.html">Record_in_channel</a> module signature. In fact, all the <code>In_channels</code> in this library satisfy this signature.</p><p>The <a href="Bio_io/Fasta/index.html">Fasta</a> and <a href="Bio_io/Fastq/index.html">Fastq</a> modules provide <code>Record</code> and <code>In_channel</code>s for reading FASTA and FASTQ files.</p><p>There are modules for reading &quot;delimited&quot; files like <a href="Bio_io/Btab/index.html">Btab</a> and <a href="Bio_io/Btab_queries/index.html">Btab_queries</a>.</p><h2 id="extending-bio_io"><a href="#extending-bio_io" class="anchor"></a>Extending Bio_io</h2><p>The <a href="Bio_io/Record_in_channel/Make/index.html">Record_in_channel.Make</a> functor can be used to make new specialized records and input channels. To do so, you need a module that satisfies the <a href="Bio_io/Record_in_channel/module-type-In_channel_input_record/index.html">In_channel_input_record</a> signature. <code>Bio_io</code> has a couple of <code>In_channel</code> types in the <a href="Bio_io/Private/index.html">Private</a> module you can use for this, and then add in the <code>input_record</code> function.</p><p>For examples, see the definitions of <a href="Bio_io/Fasta/In_channel/index.html">Fasta.In_channel</a>, <a href="Bio_io/Fastq/In_channel/index.html">Fastq.In_channel</a>, and <a href="Bio_io/Btab/In_channel/index.html">Btab.In_channel</a>.</p></div></body></html>